docs/CppInteroperability/Roadmap.md

```diff
@@ -0,0 +1,141 @@
+
+**C++, as a language:** C++ is an stubbornly un-opinionated language. There is little to no consensus on anything from naming, to generic programming, value categories, or error handling (to name a few things). This lack of opinion is in sharp contrast to Swift where there is generally one, specific way to do things. Even in the C++ Standard Library there is a rather un-standard set of API models. The un-opinionated nature of C++ means there are few, if any guarantees about a given C++ API, because the same functionality may be implemented using many different language features or programming paradigms. When consuming an arbitrary C++ API, you must be prepared for *anything*. When consuming Swift APIs, this is not the case. Swift is an opinionated language which means rather than consuming *any* API pattern you are consuming *some*, specific API pattern. Swift clearly defines patterns for naming, generic programming, value categories, error handling, etc. and it is rare that a codebase deviates from the standard practice. 
+
+**C and Objective-C interoperability:** C++ interoperability will have to build off of the previous art of C and Objective-C interoperability. Starting with the name, C++ is a superset of C, which means C++ interoperability has a great starting place to build off of. C is a simple language without many high level constructs, so importing C APIs is, relatively, simple. On the other hand, Objective-C adds some high level constructs, such as reference types and a generics model. However, unlike C++, Objective-C has strong idioms; there are only so many patterns for doing some specific thing. Further, Objective-C was able to be updated to better interoperate with Swift, the generics model is a good example of this. So, while C and Objective-C interoperability is a good starting place, the model for importing these languages cannot be taken wholesale for C++ interoperability. 
+
+## Goals
+
+Safety is a top priority. The compiler should not import APIs that would be less safe in Swift than they are in C++ and interop should strive to make imported APIs *safer* in Swift than they are in C++ by providing safe API interfaces for common, unsafe C++ API patterns (such as iterators).
+
+C++ interoperability aims to provide a clear, well-defined mapping for APIs that can be imported, and to draw a clear boundary between these APIs and APIs that cannot be imported. A user should be able to read the C++ interoperability documentation and have a good idea of both how much of their C++ API will be able to be imported in Swift and what that API will look like once it is imported. 
+
+It is a goal not to have hidden performance traps. C++ APIs should have similar performance in Swift as C++ and should be called directly when possible. Imported APIs, especially containers, should not be implicitly converted to Swift types (in contrast to the automatic Objective-C bridging for NSArray and NSString). 
```

Comment:
```suggestion
It is a goal not to have hidden performance traps. C++ APIs should have similar performance in Swift as C++ and should be called directly when possible.
```

I would leave out conclusions that imply assumptions about implementation approach.  If someone discovered a way to do that implicit conversion without hidden performance traps, surely we'd take it?

Reply:
> ```suggestion
> 
> It is a goal not to have hidden performance traps. C++ APIs should have similar performance in Swift as C++ and should be called directly when possible.
> 
> ```
> 
> 
> 
> I would leave out conclusions that imply assumptions about implementation approach.  If someone discovered a way to do that implicit conversion without hidden performance traps, surely we'd take it?

Reply:
I agree with Dave here. Avoiding performance traps is important, but if we can allow conversions in certain cases or on demand I'm not sure the goals should be against this. For instance would we want to support cases where a std::vector is explicitly assigned to or passed as a Swift Array  type (via a conversion) or does it seem like the right thing to do to have a type error?